.github/workflows/save-base-coverage.yml

```diff
@@ -0,0 +1,37 @@
+name: save base branch coverage
+
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  ubuntu:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
+        with:
+          ref: 'refs/heads/main'
+
+      - uses: actions/setup-python@8d9ed9ac5c53483de85588cdf95a591a75ab9f55
+        with:
+          python-version: '3.13'
+          cache: 'pip'
+          cache-dependency-path: |
+            '.github/workflows/requirements/unit-tests.txt'
+
+      - name: Install Python dependencies
+        run: |
+          pip install .
+          pip install -r .github/workflows/requirements/unit-tests.txt
+
+      - name: Generate coverage
+        run: |
+          coverage run -m pytest
+          coverage json -o coverage.json
+
+      - name: Upload coverage as artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: base-coverage-${{ github.sha }}
+          path: coverage.json
```

Comment:
I wonder if we can merge the logic here with `coverage.yml` so the full workflow looks like,
```
- checkout code (whether PR or main branch, e.g. always)
- set up python (always)
- install dependencies (always)
- generate pr coverage (always)
- if pr
  - attempt to download coverage for HEAD^1
  - if failed 
    - checkout HEAD^1 (probably just run git checkout HEAD^1 instead of using action)
    - generate coverage and place in expected artifact location
  - compare coverage and print pass / (failure & exit)
- if not pr
  - upload coverage artifact as SHA-1 for future PRs
```

Which would allow us to run the same workflow on main and prs and maintain less divergent code.